output/kafka/config.go

```diff
@@ -0,0 +1,305 @@
+}
+
+var compressionModes = map[string]sarama.CompressionCodec{
+	// As of sarama 1.24.1, zstd support is broken
+	// (https://github.com/Shopify/sarama/issues/1252), which needs to be
+	// addressed before we add support here.
+	"none":   sarama.CompressionNone,
+	"no":     sarama.CompressionNone,
+	"off":    sarama.CompressionNone,
+	"gzip":   sarama.CompressionGZIP,
+	"lz4":    sarama.CompressionLZ4,
+	"snappy": sarama.CompressionSnappy,
+}
+
+//const (
+//	saslTypePlaintext   = sarama.SASLTypePlaintext
+//	saslTypeSCRAMSHA256 = sarama.SASLTypeSCRAMSHA256
+//	saslTypeSCRAMSHA512 = sarama.SASLTypeSCRAMSHA512
+//)
+
+func DefaultConfig() Config {
+	return Config{
+		Hosts:              nil,
+		TLS:                nil,
+		Kerberos:           nil,
+		Timeout:            30 * time.Second,
+		BulkMaxSize:        2048,
+		BulkFlushFrequency: 0,
+		Metadata: MetaConfig{
+			Retry: MetaRetryConfig{
+				Max:     3,
+				Backoff: 250 * time.Millisecond,
+			},
+			RefreshFreq: 10 * time.Minute,
+			Full:        false,
+		},
+		KeepAlive:        0,
+		MaxMessageBytes:  nil, // use library default
+		RequiredACKs:     nil, // use library default
+		BrokerTimeout:    10 * time.Second,
+		Compression:      "gzip",
+		CompressionLevel: 4,
+		Version:          libkafka.Version("1.0.0"),
+		MaxRetries:       3,
+		Headers:          nil,
+		Backoff: BackoffConfig{
+			Init: 1 * time.Second,
+			Max:  60 * time.Second,
+		},
+		ClientID:       "elastic-agent",
+		ChanBufferSize: 256,
+		Username:       "",
+		Password:       "",
+	}
+}
+
+func (client *Config) Validate() error {
+	if client.Enabled {
+		if len(client.Hosts) == 0 {
+			return errors.New("missing required field 'hosts'")
+		}
+
+		if client.Topic == nil {
+			return errors.New("missing required field 'topic'")
+		}
+
+		if _, ok := compressionModes[strings.ToLower(client.Compression)]; !ok {
+			return fmt.Errorf("compression mode '%v' unknown", client.Compression)
+		}
+
+		if err := client.Version.Validate(); err != nil {
+			return err
+		}
+
+		if client.Username != "" && client.Password == "" {
+			return fmt.Errorf("password must be set when username is configured")
+		}
+
+	}
+	return nil
+}
+
+func newSaramaConfig(log *logp.Logger, config Config) (*sarama.Config, error) {
+	partitioner, err := makePartitioner(log, config.Partition)
+	if err != nil {
+		return nil, err
+	}
+
+	k := sarama.NewConfig()
+
+	// configure network level properties
+	timeout := config.Timeout
+	k.Net.DialTimeout = timeout
+	k.Net.ReadTimeout = timeout
+	k.Net.WriteTimeout = timeout
+	k.Net.KeepAlive = config.KeepAlive
+	k.Producer.Timeout = config.BrokerTimeout
+	k.Producer.CompressionLevel = config.CompressionLevel
+
+	tls, err := tlscommon.LoadTLSConfig(config.TLS)
+	if err != nil {
+		return nil, err
+	}
+
+	if tls != nil {
+		k.Net.TLS.Enable = true
+		k.Net.TLS.Config = tls.BuildModuleClientConfig("")
+	}
+
+	switch {
+	// TODO: Are we still planning on Kerberos to be beta?
```

Comment:
Is this a comment from you, or was this ported over from the Beats output?

Reply:
(This is a new comment) Kerberos is still beta and staying that way until we [get it working with Active Directory](https://github.com/elastic/beats/issues/33334) (which might happen before the shipper itself leaves beta but it's not on the current roadmap)

Reply:
The comment is new, but it refers to the configuration warning below - it's an open question that I will create an issue to track.

Reply:
https://github.com/elastic/elastic-agent-shipper/issues/168

Reply:
Can you reference the issue from this TODO?

Reply:
Done